src/recorder.rs

```diff
@@ -0,0 +1,283 @@
+        .level_filter(level.to_level_filter())
+        .default(debug_log)
+        .build()
+        .start();
+
+    // initialize async runtime
+    let rt = tokio::runtime::Builder::new_multi_thread()
+        .enable_all()
+        .worker_threads(1)
+        .thread_name("rezolus")
+        .build()
+        .expect("failed to launch async runtime");
+
+    // spawn logging thread
+    rt.spawn(async move {
+        loop {
+            tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+            let _ = log.flush();
+        }
+    });
+
+    ctrlc::set_handler(move || {
+        if RUNNING.load(Ordering::SeqCst) {
+            eprintln!("finalizing recording... please wait...");
+            RUNNING.store(false, Ordering::SeqCst);
+        } else {
+            eprintln!("terminating...");
+            std::process::exit(2);
+        }
+    })
+    .expect("failed to set ctrl-c handler");
+
+    // spawn recorder thread
+    rt.block_on(async move {
+        recorder(addr, destination, temporary, interval).await;
+    });
+}
+
+async fn recorder(
+    addr: SocketAddr,
+    destination: std::fs::File,
+    temporary: std::fs::File,
+    interval: Duration,
+) {
+    let mut temporary = tokio::fs::File::from_std(temporary);
+
+    let mut interval = tokio::time::interval(interval);
+
+    let mut client = None;
+
+    while RUNNING.load(Ordering::Relaxed) {
+        if client.is_none() {
+            debug!("connecting to Rezolus...");
+
+            match TcpStream::connect(addr).await {
+                Ok(s) => {
+                    if s.set_nodelay(true).is_err() {
+                        continue;
+                    }
+
+                    debug!("performing http2 handshake...");
+
+                    if let Ok((h2, connection)) = ::h2::client::handshake(s).await {
+                        tokio::spawn(async move {
+                            let _ = connection.await;
+                        });
+
+                        if let Ok(h2) = h2.ready().await {
+                            debug!("connection to Rezolus is established");
+
+                            client = Some(h2);
+                        }
+                    }
+                }
+                Err(e) => {
+                    error!("error connecting to Rezolus: {e}");
+                }
+            }
+
+            continue;
+        }
+
+        let c = client.take().unwrap();
+
+        if let Ok(mut sender) = c.clone().ready().await {
+            let request = http::request::Builder::new()
+                .version(Version::HTTP_2)
+                .method(Method::GET)
+                .uri(&format!("http://{addr}/metrics/binary"))
+                .body(())
+                .unwrap();
+
+            interval.tick().await;
+
+            let start = Instant::now();
+
+            if let Ok((response, _)) = sender.send_request(request, true) {
+                if let Ok(response) = response.await {
```

Comment:
Can we use the wait for the full body in the response here rather than manually handling it in chunks? (The reverse of the current rpc-perf change which switches the other way)

Reply:
The h2 client doesn't offer this afaik. It's more oriented towards being able to handle streaming responses.